gate/gate-web/src/test/java/com/netflix/spinnaker/gate/security/oauth/OAuth2IntegrationWithWireMockTest.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * <ul>
+ *   <li>/login/oauth/authorize → redirects to the app with code and state
+ *   <li>/login/oauth/user → returns user info JSON
+ * </ul>
+ *
+ * The application runs on a random port, and RestTemplate is configured to follow redirects. This
+ * test also verifies handling of null user ID without causing NPE.
+ */
+@SpringBootTest(
+    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
+    properties = {
+      "spring.security.oauth2.client.registration.github.client-id=client-id",
+      "spring.security.oauth2.client.registration.github.client-secret=client-secret",
+      "spring.security.oauth2.client.registration.github.scope=user,email",
+      "spring.security.oauth2.client.registration.userInfoMapping.email=email",
+      "spring.security.oauth2.client.registration.userInfoMapping.firstName=firstname",
+      "spring.security.oauth2.client.registration.userInfoMapping.lastName=name",
+      "spring.security.oauth2.client.registration.userInfoMapping.username=login"
+    })
+@TestPropertySource(properties = {"spring.config.location=classpath:gate-test.yml"})
+@Import(OAuth2TestConfiguration.class)
+public class OAuth2IntegrationWithWireMockTest {
+
+  // using restTemplate as it follows redirects which is required while testing
+  // the OAuth flows
+  @Autowired private RestTemplate restTemplate;
+
+  @LocalServerPort private int appPort;
+
+  @RegisterExtension
+  static WireMockExtension githubMockServer =
+      WireMockExtension.newInstance()
+          .options(wireMockConfig().dynamicPort().extensions(new RedirectWithStateTransformer()))
+          .build();
+
+  /** To prevent attempts to connect to clouddriver */
+  @MockBean private DefaultProviderLookupService defaultProviderLookupService;
+
+  /** To prevent attempts to connect to front50 */
+  @MockBean private ApplicationService applicationService;
+
+  /** To prevent periodic calls to service's /health endpoints */
+  @MockBean private DownstreamServicesHealthIndicator downstreamServicesHealthIndicator;
+
+  // Provide WireMock URLs into Spring properties before context initialization uses them
+  @DynamicPropertySource
+  static void dynamicProperties(DynamicPropertyRegistry registry) {
+    String base = "http://localhost:" + githubMockServer.getPort();
+    registry.add(
+        "spring.security.oauth2.client.provider.github.authorization-uri",
+        () -> base + "/login/oauth/authorize");
+    registry.add(
+        "spring.security.oauth2.client.provider.github.token-uri",
+        () -> base + "/login/oauth/access_token");
+    registry.add(
+        "spring.security.oauth2.client.provider.github.user-info-uri",
+        () -> base + "/login/oauth/user");
+  }
+
+  @Test
+  void whenOAuth2UserInfoHasNullsThenAuthenticationSucceeds() {
+    // Now appPort (@LocalServerPort) is available — set the supplier so transformer can use it
+    RedirectWithStateTransformer.setAppPortSupplier(() -> appPort);
```

Comment:
Could this move to a `@BeforeAll` so it only happens once?

Reply:
If the port is available by then, I will try.

Reply:
done